addons/delivery/models/sale_order.py

```diff
@@ -111,8 +111,10 @@ def _create_delivery_line(self, carrier, price_unit):
             values['name'] += '\n' + 'Free Shipping'
         if self.order_line:
             values['sequence'] = self.order_line[-1].sequence + 1
-        if self.picking_ids:
-            self.picking_ids.write({'carrier_id': carrier.id})
+
+        to_update = self.picking_ids.filtered(lambda p: p.state not in ['cancel', 'done'] and p.carrier_tracking_ref == False)
```

Comment:
I am not sure of this filter for two reasons:
1: if you have set your integration level to 'rate only', it means that you create the label after having the picking done. So you may want to have the propagation of the carrier on the pickings even if they are done.
2: So if you alllow to write it when done, you may not know if the carrier will generate a tracking ref, as if you have computed prices or fixed charges, it won't create a ref.

So maybe the condition should be something like "p.state not in ('done', 'cancel') or (p.state == 'done' and (p.carrier.delivery_type not in ('fixed', 'base_on_rule') and not p.carrier_tracking_ref)

Does it looks good for you. It means that it'll change the carrier, in done if no traking reference has been made  for it and it is a .
There are no case where we are not in done and having a reference tracking number, because generation of label needs picking to be done.

Reply:
@pimodoo 
You have a point and I checked on the code based on your input. Thx for a constructive review and input ❤️ 

So yes, the carrier will be set only if the state is not in done and the method will be called based on `_send_shipper` to create a NEW sale order delivery line or a zero sale order delivery line triggered by a picking vs. created based on the sale order wizard which was my entry point and why we should not overwrite on done or already processed in terms of a tracking reference was set. So, we really want to overwrite just for planning of not yet processed via the sale order.

Not sure why the carrier should be switched after the picking was confirmed/done?! Even if the label printing and the final cost communication is done at this point...the carrier or the rate is based on an carrier already set and therefore should not be changed, right?

Not sure if I am missing something, but for me the domain is as precise as we can get it...
Maybe you can check yourself again and let me know your further thoughts...

I wished this kind of review would have been done internally with bug fixing, as now nobody is paying me for the work which should have been done by the multi national Odoo S.A. and its hundreds of developers 😉
But hey, as long as it gets fixed and I get appreciation and respect (at least on a personal level), it is fine for me if you guys improve the "monster" where I and a lot of others point for quite a long time...